pyroaring/abstract_bitmap.pxi

```diff
@@ -209,44 +209,85 @@ cdef class AbstractBitMap:
         """
         return self.from_ptr(croaring.roaring_bitmap_flip(self._c_bitmap, start, end))
 
-    @classmethod
-    def union(cls, *bitmaps):
+    def copy(self):
+        """
+        Return a copy of a set.
+        """
+        return self.__class__(self, copy_on_write=self._c_bitmap.copy_on_write)
+
+    def isdisjoint(self, other):
+        """
+        Return True if two sets have a null intersection.
+        """
+        return AbstractBitMap.intersection_cardinality(self, other) == 0
+
+    def issubset(self, other):
+        """
+        Report whether another set contains this set.
+        """
+        return AbstractBitMap.intersection_cardinality(self, other) == len(self)
+
+    def issuperset(self, other):
+        """
+        Report whether this set contains another set.
+        """
+        return AbstractBitMap.intersection_cardinality(self, other) == len(other)
+
+    def difference(self, *others):
+        """
+        Return the difference of two or more sets as a new set.
+
+        (i.e. all elements that are in this set but not the others.)
+        """
+        difference = BitMap(self)
+        for other in others:
+            difference.__ixor__(difference & other)
+
+        if not isinstance(self, BitMap):
+            return self.__class__(difference)
+        return difference
+
+
+    def symmetric_difference(self, other):
+        """
+        Return the symmetric difference of two sets as a new set.
+
+        (i.e. all elements that are in exactly one of the sets.)
+        """
+        return AbstractBitMap.difference(
+            AbstractBitMap.union(self, other),
+            AbstractBitMap.intersection(self, other)
+        )
+
+    def union(self, *others):
         """
         Return the union of the bitmaps.
 
         >>> BitMap.union(BitMap([3, 12]), BitMap([5]), BitMap([0, 10, 12]))
         BitMap([0, 3, 5, 10, 12])
         """
-        size = len(bitmaps)
-        cdef croaring.roaring_bitmap_t *result
-        cdef AbstractBitMap bm
-        cdef vector[const croaring.roaring_bitmap_t*] buff
-        if size <= 1:
-            return cls(*bitmaps)
-        else:
-            for bm in bitmaps:
-                bitmaps[0].__check_compatibility(bm)
-                buff.push_back(bm._c_bitmap)
-            result = croaring.roaring_bitmap_or_many(size, &buff[0])
-            return (<AbstractBitMap>cls()).from_ptr(result) # FIXME to change when from_ptr is a classmethod
+        union = BitMap(self)
+        for other in others:
+            union.__ior__(other)
```

Comment:
I would be surprised that this new implementation is really faster than the original one.

Here is a first quick benchmark:
```bash
python -m timeit -s 'from pyroaring import BitMap; import random; size=1000000; bitmaps=[BitMap(random.sample(range(size), size//20)) for _ in range(100)]' 'BitMap.union(*bitmaps)'
```
* new implementation: 100 loops, best of 3: 3.89 msec per loop
* old implementation: 100 loops, best of 3: 2.87 msec per loop

Reply:
Yeah, it looks like it depends on the dataset. I tested on much more dense sets.

```python
import random
size = 1000000
bitmaps = []
for i in range(100):
    b = BitMap(range(size))
    for i in range(size // 20):
        b.discard(random.randint(0, size))
    b.run_optimize()
    bitmaps.append(b)
```
* new implementation: 239 µs ± 19.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
* old implementation: 480 µs ± 17.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

And the difference is much more drastic without `run_optimize`.

* new implementation: 1.33 ms ± 52.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
* old implementation: 25.3 ms ± 1.79 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)